### PR TITLE
Add new ledger event exposing drep registration/unregistration

### PR DESCRIPTION
**Pull request by smelc**
# Description

These new events will make it possible to write tests in [cardano-testnet](https://github.com/IntersectMBO/cardano-node/tree/master/cardano-testnet) that do not rely on timeouts.

# Checklist

- [X] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- NA When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- NA The version bounds in `.cabal` files for all affected packages are updated.
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- NA Cabal files are formatted (use `scripts/cabal-format.sh`)
- NA [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [X] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs [hunk 1]
```diff
@@ -126,7 +127,9 @@ instance
         pure (2, ConwayCommitteeHasPreviouslyResigned keyH)
       k -> invalidKey k
 
-newtype ConwayGovCertEvent era = GovCertEvent (Event (EraRule "GOVCERT" era))
+data ConwayGovCertEvent c =
+    DRepRegistration !(Credential 'DRepRole c)
+  | DRepUnregistration !(Credential 'DRepRole c)
```
Comment (Jimbo4350) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs]: @smelc lets also add events for hot key authorization and drep update

Comment (smelc) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs]: @Jimbo4350> done :+1:

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs [hunk 2]
```diff
@@ -126,7 +129,17 @@ instance
...
+  = -- | Event indicating that a DRep is registrating
+    DRepRegistration !(Credential 'DRepRole c)
+  | -- | Event indicating that a DRep is unregistrating
+    DRepUnregistration !(Credential 'DRepRole c)
+  | -- | Event indicating that the hot key (corresponding to the given cold key) is being overwritten
+    OverwriteCommitteeHotKey !(Credential 'ColdCommitteeRole c) !(Credential 'HotCommitteeRole c)
+  | -- | Event indicating that a cold key is being resigned
+    ResignCommitteeColdKey !(Credential 'ColdCommitteeRole c)
+  | -- | Event indicating that a drep expiry and anchor are updated
+    DRepUpdate !(Credential 'DRepRole c) !(StrictMaybe (Anchor c))
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/GovCert.hs:142]: Why not?
```suggestion
newtype ConwayGovCertEvent era
  = ConwayGovCertEvent (ConwayGovCert (EraCrypto era))
```

**Review by lehins — CHANGES_REQUESTED** [review]
Guys, this is counterproductive. Emitting an event for applied certificate is not what events where designed for.

You are essentially emitting a `Signal rule`, but signal is the input to the rule.

So you are literally writing a function that returns its argument through a complex mechanism of events.

**Comment by lehins** [comment]
Closing in favor of a new testing approach that @Jimbo4350 is working on.

**Pull request closed without merging**